### PR TITLE
[ME-3065] CLI/Connector Support for Kubernetes Sockets

### PR DESCRIPTION
**Pull request by adrianosela**
## [[ME-3065](https://mysocket.atlassian.net/browse/ME-3065)] CLI/Connector Support for Kubernetes Sockets

CLI and initial connector support for Kubernetes sockets. Note that the feature is still, for all intents and purposes, unreleased.

We have limited support for upstream authentication methods (only support auth with provided api server + client cert + client cert key). Even the schema of the upstream configuration is likely going to change in the next couple of weeks. The work to support more upstream authentication methods e.g. kubeconfig path, specific kubeconfig context, etc. is tracked internally via https://mysocket.atlassian.net/browse/ME-3107

Note that the border0 proxy performs authentication and authorization based on border0 policies. The plan is to have fine grained authorization (based on border0 policies) in the connector. That is still to-be-built (but right now there is no fine grained actions in the policy so it doesn't matter...). That work (to build authz in the connector) is tracked internally via https://mysocket.atlassian.net/browse/ME-3106

Fixes # (issue)
- https://mysocket.atlassian.net/browse/ME-3065

## Type of change

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Plenty of manual testing done against a local [kind](https://kind.sigs.k8s.io/) cluster.

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code

[ME-3065]: https://mysocket.atlassian.net/browse/ME-3065?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

- **New Features**
  - Introduced functionality to configure local Kubernetes kubeconfig via the CLI.
  - Added commands for setting and getting Kubernetes configurations.
  - Enhanced support for managing Kubernetes sockets, including API proxy support and access control.
  
- **Bug Fixes**
  - Improved handling and validation of Kubernetes service configurations.
  - Enhanced error handling and validation for Kubernetes API requests.

- **Refactor**
  - Updated internal structures to include Kubernetes-related settings for better integration.

- **Tests**
  - Implemented tests for Kubernetes request handling and conversions.

### CODE REVIEW

File: internal/connector_v2/service.go [hunk 1]
```diff
@@ -852,6 +862,10 @@ func (c *ConnectorService) Listen(socket *border0.Socket) {
 		if err := sqlauthproxy.Serve(l, *handlerConfig); err != nil {
 			logger.Error("sql proxy failed", zap.String("socket", socket.SocketID), zap.Error(err))
 		}
+	case k8sapiProxyConfig != nil:
+		if err = k8sapilib.Serve(l, k8sapiProxyConfig); err != nil {
```
Comment (btoonk) [inline internal/connector_v2/service.go:866]: I guess it's better to use `if err := `

File: internal/k8sapilib/server.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	// More info:
+	// - There's a generic open issue with Go upstream: https://github.com/golang/go/issues/65930
+	// - There's a specific issue with Go upstream: https://github.com/golang/go/issues/67152
+	// - The problematic code is in https://go.googlesource.com/crypto/+/master/ssh/tcpip.go
+	// - There's a fix that works for us in https://go-review.googlesource.com/c/crypto/+/562756
+
+	dummyListenerFilepath := filepath.Join(os.TempDir(), strconv.FormatInt(time.Now().UnixNano(), 10))
+	defer os.Remove(dummyListenerFilepath)
+
+	dummyListener, err := net.Listen("unix", dummyListenerFilepath)
```
Comment (btoonk) [inline internal/k8sapilib/server.go:69]: This won't work on windows. That is fine for now, but just so you know.

File: internal/k8sapilib/handler.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+		)
+
+		// filter out offending requests
+		if err := evaluateAccess(logger, socket, r); err != nil {
+			http.Error(w, err.Error(), http.StatusForbidden)
+			return
+		}
+
+		// determine target kubernetes api settings
+		targetSettings, err := getKubernetesAPISettings(socket)
```
Comment (btoonk) [inline internal/k8sapilib/handler.go:36]: can we initialise this, so we won't have to this on each request.

File: internal/k8sapilib/handler.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	return &models.KubernetesAPISettings{
+		Server:                   kubeConfig.Host,
+		CertificateAuthorityData: base64.StdEncoding.EncodeToString(kubeConfig.CAData),
+		ClientCertificateData:    base64.StdEncoding.EncodeToString(kubeConfig.CertData),
+		ClientKeyData:            base64.StdEncoding.EncodeToString(kubeConfig.KeyData),
+	}, nil
+}
+
+func buildKubernetesProxyTransport(caCert, clientCert, clientKey string) (*http.Transport, error) {
+	tlsConfig := &tls.Config{InsecureSkipVerify: false}
```
Comment (btoonk) [inline internal/k8sapilib/handler.go:95]: this is the default value, so you can just do` tlsConfg := &tls.Config{}`

File: internal/k8sapilib/handler.go (hunk 3 above)
Comment (adrianosela) [reply internal/k8sapilib/handler.go:36]: yes, also doing `getKubernetesProxy` just once per socket now.

File: internal/k8sapilib/handler.go (hunk 4 above)
Comment (adrianosela) [reply internal/k8sapilib/handler.go:95]: i left it in for explicitness, but removed.

**Comment by adrianosela** [comment]
> This wil only work in connector mode, not with `socket connect`

To be tracked internally via https://mysocket.atlassian.net/browse/ME-3146

File: internal/k8sapilib/server.go (hunk 2 above)
Comment (adrianosela) [reply internal/k8sapilib/server.go:69]: To be tracked via https://mysocket.atlassian.net/browse/ME-3147

**Review by btoonk — APPROVED** [review]

**Pull request merged**